protocol/contracts/templegold/TempleGoldStaking.sol

```diff
@@ -0,0 +1,489 @@
+        emit RewardDistributionCoolDownSet(_cooldown);
+    }
+
+    /**
+     * @notice Set half time parameter for calculating vote weight.
+     * @dev The voting half-time variable determines the time it takes until half the voting weight is reached for a stake.
+     *      Formular from st-yETH https://docs.yearn.fi/getting-started/products/yeth/overview
+     * @param _halfTime Cooldown in seconds
+     */
+    function setHalfTime(uint256 _halfTime) external override onlyElevatedAccess {
+        if (_halfTime == 0) { revert CommonEventsAndErrors.ExpectedNonZero(); }
+        halfTime = _halfTime;
+        emit HalfTimeSet(_halfTime);
+    }
+
+    /**
+      * @notice For migrations to a new staking contract if TGLD changes
+      *         1. Withdraw `staker`s tokens to the new staking contract (the migrator)
+      *         2. Any existing rewards are claimed and sent directly to the `staker`
+      * @dev Called only from the new staking contract (the migrator).
+      *      `setMigrator(new_staking_contract)` needs to be called first
+      * @param staker The staker who is being migrated to a new staking contract.
+      * @param amount The amount to migrate - generally this would be the staker's balance
+      */
+    function migrateWithdraw(address staker, uint256 amount) external override onlyMigrator {
+        _withdrawFor(staker, msg.sender, amount, true, staker);
+    }
+    
+    /**
+     * @notice Distributed TGLD rewards minted to this contract to stakers
+     * @dev This starts another 7-day rewards distribution. Calculates new `rewardRate` from any left over rewards up until now
+     */
+    function distributeRewards() external {
+        if (distributionStarter != address(0) && msg.sender != distributionStarter) { revert CommonEventsAndErrors.InvalidAccess(); }
+        // Mint and distribute TGLD if no cooldown set
+        if (lastRewardNotificationTimestamp > 0 && 
+            lastRewardNotificationTimestamp + rewardDistributionCoolDown > block.timestamp) 
+                { revert CannotDistribute(); }
+        _distributeGold();
+        uint256 rewardAmount = nextRewardAmount;
+        if (rewardAmount == 0 ) { revert CommonEventsAndErrors.ExpectedNonZero(); }
+        nextRewardAmount = 0;
+        _notifyReward(rewardAmount);
+    }
+
+    /**
+     * @notice Stake
+     * @param amount Amount of staking token
+     */
+    function stake(uint256 amount) external override {
+        stakeFor(msg.sender, amount);
+    }
+
+    /**
+     * @notice Stake all balance of staker
+     */
+    function stakeAll() external override {
+        uint256 balance = stakingToken.balanceOf(msg.sender);
+        stakeFor(msg.sender, balance);
+    }
+
+    /**
+     * @notice Stake for account when contract is not paused.
+     * @param _for Account to stake for
+     * @param _amount Amount of staking token
+     */
+    function stakeFor(address _for, uint256 _amount) public override whenNotPaused {
```

Comment:
I'm not sure I've ever actually tested this -- do modifiers (`whenNotPaused`) get called when its called from another internal function like from `stake()`?

Reply:
I think you mean "when called from another external function", `stake()` is external.  yes they do